### ISSUE TITLE
[BUG] Race condition in upward scrolling compensation system could fire scrollBy when user is scrolling

### ISSUE THREAD

**Issue by khovansky-al** [petyosi/react-virtuoso#605]
**Describe the bug**
Upward scrolling compensation might kick in while the user is still scrolling. This is noticeable on lower-end devices and when the components might take time to mount.

**Reproduction**
Reproduction is tricky. I made a demo with variable-height elements to ensure there is a decent amount of compensation and added some synchronous calculations to the render function to slow the components down.
Originally noticed in Firefox on Windows while testing lower-end devices, confirmed in a VM.
https://healthy-thunder-sunflower.glitch.me (using glitch.me because I needed to compare it to a github fork).
Depending on how fast your computer is, you might want to increase the value in the component to slow things more though can't guarantee it would help. Almost impossible to repro on macOS Firefox. Definitely not as obvious as in the video, no matter the slowdown.

Looking at the [code](https://github.com/petyosi/react-virtuoso/blob/master/src/upwardScrollFixSystem.ts#L88-L90), deviation pipe filters out new values that have `isScrolling === true`. The question is why does `isScrolling` become `false` during scrolling.

Briefly looking at [stateFlagsSystem](https://github.com/petyosi/react-virtuoso/blob/13331d527472eb835bb0d73092bf8444e0bfb000/src/stateFlagsSystem.ts#L62-L67), scroll end is detected by reporting `false` in a debounced handler. I suppose task queue could be full and let the debounced `false` value go through before a new scroll event is processed. In other words, before a new `scrollTop` value triggers the `true` branch (and resets the `false`, I guess, not sure how exactly `merge` works here).

Another finding: if the filter predicate passes, the invocation is [throttled](https://github.com/virtuoso-dev/urx/blob/35c5af30cb8d322a42e3c46a381e5f2667261096/packages/urx/src/pipe.ts#L255-L271), which schedules a new task using `setTimeout`. Following the same logic of tasks scheduling in the event loop, the actual value of `isScrolling` could be different by the time the subscription callback is fired.

Attempted to make an illustration of how event loop might look.
![virtuoso_diagram_1](https://user-images.githubusercontent.com/4955642/157633463-58b43b7a-d956-4ce7-981f-21c4229b36fc.png)

**To Reproduce**
1. Go to the provided demo
2. Scroll up
3. Observe viewport jumping down against the scroll sometimes

It should be more noticeable if you try to accumulate more deviation by scrolling non-stop.
If you open devtools on the side, you could see that `margin-top` would  reset to 0 sometimes during scrolling.
The amount of difference with each element could change depending on the first item, so sometimes refreshing could give a more janky collection.

<video src=https://user-images.githubusercontent.com/4955642/157632286-b8de2cee-29e1-4322-b2dc-4f351375a327.mp4></video>

Recorded on FF 98, Windows 11 ARM VM. Used touchpad for scrolling.

Btw applying a fix that only checks for `isScrolling` before `publish`, without correcting for available `scrollTop`, exposes accumulated `margin-top` if you happen to reach the top.

**Expected behavior**
Compensation should be firing only after user truly stopped scrolling

**Desktop (please complete the following information):**
Firefox, tested on v97+ on Windows.

**Comment by petyosi** [comment]
Thanks for the time you took to investigate this. I think you're on the right track for a possible improvement here. It might be a faulty leakage of isScrolling=false due to the way I've combined throttling. I will take a look.

**Comment by petyosi** [comment]
@khovansky-al you've mentioned a fix that you tried. Can you point me to it?

**Comment by khovansky-al** [comment]
I tried checking for `isScrolling` before publishing `scrollBy` and it seemed to have stabilized things.
That's a dirty research branch so it is rigged with extra log statements and also includes stuff for another problem with stale values that I've yet to report here. Might be hard to read, sorry about that.

https://github.com/khovansky-al/react-virtuoso/commit/458b195160752f18d9b15196c8751766af74269e#diff-37671eeeb5c5f19fe3e4f4d0dc88569ffe60e88e3af45103aaa59589efecc161R103

But that only addresses the second throttling issue (if the value was updated between `filter` and the subscription callback). Theoretically there's no guarantee a new scroll event would get processed before the subscription task. Ideally we'd want `isScrolling` to report values correctly.

Also as mentioned, with addition of this fix, you are able to scroll all the way to the top without flushing deviation. This exposes the compensation margin, so I added a (very) quick fix for that too.
https://github.com/khovansky-al/react-virtuoso/commit/458b195160752f18d9b15196c8751766af74269e#diff-37671eeeb5c5f19fe3e4f4d0dc88569ffe60e88e3af45103aaa59589efecc161R59-R61

Sandbox with forked version: https://virtuoso-forked-fix.glitch.me.

Thanks for looking into this.

**Comment by petyosi** [comment]
Agreed on the `isScrolling` - if that's the root cause, that should be fixed. I will try to reproduce it. I will double-check on that, hopefully over the weekend.

**Comment by petyosi** [comment]
Breaking the problem further and based on your diagram, I tried to cause debounce leak due to heavy queue. [Here's a simplified test](https://github.com/virtuoso-dev/urx/commit/d72db19453900646362597d37e1e35ada98a72da). Maybe I am missing something here, but it looks like the extra ops don't affect it. But I might be too gentle. 

Ultimately, if the rendered components can cause sufficient timer delay, `isScrolling` is not the only flag that is going to get miscalculated. There are several other timers in the codebase. Here are a few ideas that can help:

- Can you create a reliably failing e2e test? Ideally, without heavy components, as it gets quite subjective. 
- Can you try to ease the rendering through the scroll seek placeholders? They are meant to solve exactly that.
